plugins/logger/filesystem_logger.cpp

```diff
@@ -32,35 +34,179 @@ FLAG(string,
 /// Legacy, backward compatible "osquery_log_dir" CLI option.
 FLAG_ALIAS(std::string, osquery_log_dir, logger_path);
 
+FLAG(bool, logger_rotate, false, "Use filesystem log rotation");
+FLAG(uint64,
+     logger_rotate_max_files,
+     25,
+     "Max number of files to keep in rotation");
+FLAG(uint64,
+     logger_rotate_size,
+     25 * 1024 * 1024,
+     "Size for each filesystem log in bytes");
+
 FLAG(int32, logger_mode, 0640, "Decimal mode for log files (default '0640')");
 
 const std::string kFilesystemLoggerFilename = "osqueryd.results.log";
 const std::string kFilesystemLoggerSnapshots = "osqueryd.snapshots.log";
 
+bool LogRotate::shouldRotate() {
+  if (this->fileSize(path_) >= this->getRotateSize()) {
+    return true;
+  }
+  return false;
+}
+
+Status LogRotate::rotate(size_t max_files) {
+  if (!pathExists(path_)) {
+    return Status::failure("Path under rotation does not exist");
+  }
+  size_t offset = 1;
+  std::vector<std::string> moves{getRotateFile(offset)};
+  while (this->pathExists(moves.back())) {
+    moves.push_back(getRotateFile(++offset));
+  }
+
+  auto s = this->moveFiles(moves, max_files);
+  if (!s.ok()) {
+    return s;
+  }
+  return this->moveFile(path_, *moves.begin());
+}
+
+size_t LogRotate::getRotateSize() {
+  return FLAGS_logger_rotate_size;
+}
+
+size_t LogRotate::fileSize(const std::string& filepath) {
+  boost::system::error_code ec;
+  auto size = fs::file_size(filepath, ec);
+  return (ec) ? 0 : size;
+}
+
+bool LogRotate::pathExists(const std::string& path) {
+  return osquery::pathExists(path).ok();
+}
+
+Status LogRotate::removeFile(const std::string& path) {
+  return osquery::removePath(path);
+}
+
+Status LogRotate::moveFile(const std::string& source, const std::string& dest) {
+  return osquery::movePath(source, dest);
+}
+
+Status LogRotate::compressFile(const std::string& source,
+                               const std::string& dest) {
+  return osquery::compress(source, dest);
```

Comment:
compression can be pretty slow. Is this in a blocking path?

Reply:
Yeap it runs inline. We don't have a great way of managing short-running threads right now but I am happy to follow up and propose a safe solution. Running this async is a little beyond the initial v1 I was aiming for and since this is a boutique feature I didn't want to over-engineer.